### PR TITLE
continue `ParamEnv` to `TypingEnv` transition

### PR DESCRIPTION
**Pull request by lcnr**
cc #132279

r? @compiler-errors

### CODE REVIEW

**Comment by lcnr** [comment]
cc @BoxyUwU on the last commit

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1817,9 +1817,7 @@ pub(crate) fn clean_ty<'tcx>(ty: &hir::Ty<'tcx>, cx: &mut DocContext<'tcx>) -> T
                     let ct = if let hir::ConstArgKind::Anon(hir::AnonConst { def_id, .. }) =
                         const_arg.kind
                     {
-                        // Only anon consts can implicitly capture params.
-                        // FIXME: is this correct behavior?
-                        let typing_env = ty::TypingEnv::from_param_env(cx.tcx.param_env(*def_id));
+                        let typing_env = ty::TypingEnv::post_analysis(cx.tcx, *def_id);
                         cx.tcx.normalize_erasing_regions(typing_env, ct)
```
Comment (lcnr) [inline src/librustdoc/clean/mod.rs:1821]: we're using `normalize_erasing_regions` to evaluate the constant, so it's gonna switch to `Reveal::All` anyways, cc @BoxyUwU

**Comment by RalfJung** [comment]
What about the part where `InterpCx` stores a `TypingEnv`?

Currently the `InterpCx` setup in Miri still seems rather odd to me.

**Comment by RalfJung** [comment]
Separate PR is fine, I was just confused since you didn't mention it in the PR description. :)

**Review by compiler-errors — APPROVED** [review]
r=me when green

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 2]
```diff
@@ -36,8 +38,9 @@ pub struct InterpCx<'tcx, M: Machine<'tcx>> {
     /// we are evaluating (if this is CTFE).
     pub tcx: TyCtxtAt<'tcx>,
 
-    /// Bounds in scope for polymorphic evaluations.
-    pub(crate) param_env: ty::ParamEnv<'tcx>,
+    /// The current context in case we're evaluating in a
+    /// polymorphic context. This always uses `ty::TypingMode::PostAnalysis`
+    pub typing_env: ty::TypingEnv<'tcx>,
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:43]: I don't think we should make this public, I want to be sure nobody mutates this field... I'll make a PR to privatize this again.

**Comment by lcnr** [comment]
Passing `TypingEnv` to the const-eval queries seems to have a bigger perf impact. Will look into reverting this change and making a new `TypingEnv` from scratch in the query

**Pull request merged**